README.md

# CustomWillPopScope

[![pub package](https://img.shields.io/pub/v/custom_will_pop_scope.svg)](https://pub.dev/packages/custom_will_pop_scope)

>  _Note: This package can be used on any platform, and is not specific to iOS._

  
<br/>

  <img src="https://github.com/HelmyBc/custom_will_pop_scope/assets/80924075/3a19a11e-b0d5-4f49-98f3-53ae069c3f5b" height="400"/>


## CustomWillPopScope

  

When using this widget, be sure to update `canReturn` based on the state of your screen, and only set it to `false` when the screen **should not be allowed to pop**. Additionally, the `onWillPop` property should not be changed once set.

  

-------

  

A working app using this package can be found in the [example](example/lib/main.dart) folder.

  

## Usage

  

To use this package, add `custom_will_pop_scope` as a [dependency in your pubspec.yaml file](https://flutter.io/using-packages/).

  
  

## Example

### Import the library

  

``` dart

// main.dart
import  'package:decorated_icon/custom_will_pop_scope.dart';

```

### Wrap your screen

Using the included `CustomWillPopScope` widget, wrap your screen and define an `onWillPop` callback for it, and `onPopAction` if you have a function to call when the screen pops (Like removing a route name in the [example](example/lib/main.dart) ).

>  ##### Note that `onWillPop` should always return a `bool`. See the [Flutter Docs](https://api.flutter.dev/flutter/widgets/WillPopScope-class.html) for more.

  

##### The first Screen:
```dart

// first_screen.dart
class  HomeScreen  extends  StatelessWidget {
  void  navigateToNext() {
  locator<NavigationService>().navigateTo(secondScreen);
  }

  @override
  Widget  build(BuildContext  context) {
    return  Scaffold(
      appBar:AppBar(title: Text('CustomWillPopScopeDemo')),
      body:  Center(
        child:  ElevatedButton(
          onPressed: () =>  navigateToNext(),	
          child:  Padding(
            padding:  EdgeInsets.symmetric(horizontal:  24.0,vertical:  12.0),
            child:  Text('Go to second screen'),
            ),
          ),
        ),
      );
    }
}
```

  
##### The second screen (Using CustomWillPopScope):

```dart

// second_screen.dart
class  SecondScreen  extends  StatefulWidget {
  @override
  _SecondScreenState  createState() =>  _SecondScreenState();
}

  

class  _SecondScreenState  extends  State<SecondScreen> {

  final  NavigationService  _navigationService  =  locator<NavigationService>();
  /// Holds the state of the screen.
  bool  _canReturn  =  true;
  /// Shows an alert and returns `false` when `_canReturn` is `true`.
  /// This prevents the navigator from popping this route.
  Future<bool> _onWillPop() async {
    if (!_canReturn) {
    // Show an alert before returning `false`.
      showDialog(context:  context,
        builder: (context) =>  AlertDialog(
          content:  Text('Back navigation is disabled.'),
        ),
      );
      // Return `false` to prevent the route from popping.
      return  false;
    } else {
      _navigationService.removeLastRouteName();
      return  true;
    }
  }

  void  _onPopAction() {
    _navigationService.removeLastRouteName();
  }

  /// Updates `_hasChanges` with the provided value.
  void  _updateChanges(bool  value) =>  setState(() =>  _canReturn  =  value);

  @override
  Widget  build(BuildContext  context) {
    return  CustomWillPopScope(
      // canReturn is optional, it is set to true by default
      canReturn:  _canReturn,
      onWillPop:  _onWillPop,
      onPopAction:  _onPopAction,
      child:  Scaffold(
        appBar:  AppBar(title:  Text('Second Screen')),
        body:  Center(
          child:  Container(
            decoration:  BoxDecoration(
              border:  Border.all(color:  _color),
              borderRadius:  BorderRadius.circular(6.0),
            ),
            margin:  EdgeInsets.symmetric(horizontal:  12.0),
            child:  SwitchListTile(
              activeColor:  _color,
              onChanged:  _updateChanges,
              title:  Text('Enable back navigation'),
              value:  _canReturn,
            ),
          ),
        ),
      ),
    );
  }
}

```

  



CHANGELOG.md
## [1.0.0]
- Initial release.
## [2.0.0]
- Fix the inabilty to pop when there is no changes.
- Remove the `CupertinoPageTransitionBuilder` widget.
## [2.0.1]
- Update README.
- Update the Example.
## [2.0.2]
- Make `onPopAction` nullable.
## [2.0.3]
- Make the package compatible with Dart SDK >= 2.19.0
- Add documentation.
- Update README.
## [2.1.0]
- `swipeOnAndroid` added to choose whether the user can use swipe to return in android devices



example/lib/main.dart
import 'package:custom_will_pop_scope/custom_will_pop_scope.dart';
import 'package:example/locator.dart';
import 'package:example/navigation_service.dart';
import 'package:example/router.dart' as router;
import 'package:example/router.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  final theme = ThemeData(
    primarySwatch: Colors.blue,
  );

  @override
  Widget build(BuildContext context) {
    setupLocator();
    return MaterialApp(
      theme: theme,
      home: HomeScreen(),
      initialRoute: router.homeScreen,
      onGenerateRoute: router.generateRoute,
      navigatorKey: locator<NavigationService>().navigationKey,
    );
  }
}

class HomeScreen extends StatelessWidget {
  void navigateToNext() {
    locator<NavigationService>().navigateTo(secondScreen);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('CustomWillPopScope Demo')),
      body: Center(
        child: ElevatedButton(
          child: Padding(
            padding: EdgeInsets.symmetric(horizontal: 24.0, vertical: 12.0),
            child: Text('Go to second screen'),
          ),
          onPressed: () => navigateToNext(),
        ),
      ),
    );
  }
}

class SecondScreen extends StatefulWidget {
  @override
  _SecondScreenState createState() => _SecondScreenState();
}

class _SecondScreenState extends State<SecondScreen> {
  final Color _color = Colors.pink;
  final NavigationService _navigationService = locator<NavigationService>();

  /// Holds the state of the screen.
  bool _canReturn = true;

  /// Shows an alert and returns `false` when `_canReturn` is `true`.
  /// This prevents the navigator from popping this route.
  Future<bool> _onWillPop() async {
    if (!_canReturn) {
      // // Show an alert before returning `false`.
      showDialog(
        context: context,
        builder: (context) => AlertDialog(
          content: Text('Back navigation is disabled.'),
        ),
      );
      // Return `false` to prevent the route from popping.
      return false;
    } else {
      _navigationService.removeLastRouteName();
      return true;
    }
  }

  void _onPopAction() {
    _navigationService.removeLastRouteName();
  }

  /// Updates `_canReturn` with the provided value.
  void _updateChanges(bool value) => setState(() => _canReturn = value);

  @override
  Widget build(BuildContext context) {
    return CustomWillPopScope(
      // canReturn is optional, it is set to true by default
      canReturn: _canReturn,
      onWillPop: _onWillPop,
      // onPopAction is optional, use it when you have a function to call when dissmissing the enclosing [ModalRoute].
      onPopAction: _onPopAction,
      child: Scaffold(
        appBar: AppBar(title: Text('Second Screen'), backgroundColor: _color),
        body: Center(
          child: Container(
            child: SwitchListTile(
              activeColor: _color,
              onChanged: _updateChanges,
              title: Text('Enable back navigation'),
              value: _canReturn,
            ),
            decoration: BoxDecoration(
              border: Border.all(color: _color),
              borderRadius: BorderRadius.circular(6.0),
            ),
            margin: EdgeInsets.symmetric(horizontal: 12.0),
          ),
        ),
      ),
    );
  }
}
